src/SqlPersistence/Subscription/SubscriptionPersister.cs

```diff
@@ -70,55 +89,59 @@ public SubscriptionPersister(Func<DbConnection> connectionBuilder, string tableP
         }
     }
 
-    public async Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes, ContextBag context)
+    public Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageHierarchy, ContextBag context)
     {
-        var types = messageTypes.ToList();
+        var types = messageHierarchy.ToList();
 
         if (cacheFor == null)
         {
-            return await GetSubscriptions(types).ConfigureAwait(false);
+            return GetSubscriptions(types);
         }
 
-        var typeNames = types.Select(_ => _.TypeName);
-        var key = string.Join(",", typeNames);
-        CacheItem cacheItem;
-        if (cache.TryGetValue(key, out cacheItem))
-        {
-            if (DateTimeOffset.UtcNow - cacheItem.Stored < cacheFor)
-            {
-                return cacheItem.Subscribers;
-            }
-        }
+        var key = GetKey(types);
 
-        var baseSubscribers = await GetSubscriptions(types)
-            .ConfigureAwait(false);
+        var cacheItem = Cache.GetOrAdd(key,
+            valueFactory: _ => new CacheItem
+            {
+                Stored = DateTime.UtcNow,
+                Subscribers = GetSubscriptions(types)
```

Comment:
Took me a few seconds to get this but overall a nice trickery! You are basically kicking of the GetSubscriptions task and when the cacheItem.Subscribers it might already be completed and then the await is almost free from the core perspective